### DISCUSSION TITLE
Distance functions

### DISCUSSION THREAD

**Discussion by hugo-pires** [TDAmeritrade/stumpy#543] [Q&A] [answered]
Hello and congratulations for your work

Is it possible to use other distance than the euclidean one? And, in your opinion, would it make any sense?

Thank you

**Comment by seanlaw** [comment, marked as answer]
@hugo-pires Thank you for your question and welcome to the STUMPY community? Can you please elaborate on what "other distances" you are interested in using?

Natively, STUMPY supports z-normalized Euclidean distance and, when you set `normalize=False`, this changes it to "non-normalized Euclidean distance". The reason why this is fast is because there is special relationship between each distance along the pairwise distance matrix (i.e., `D[i, j]`) and its nearest diagonal neighbor (i.e., `D[i-1, j-1]`). Please see [Section IV A of this Matrix Profile II paper](https://www.cs.ucr.edu/~eamonn/STOMP_GPU_final_submission_camera_ready.pdf) for more information. 

Now, regarding your question about whether or not other distances "make any sense", that really depends on your objective. In z-normalized Euclidean distances, the matrix profile is really finding the one nearest neighbor subsequence with the same shape (i.e., it doesn't depend on amplitude) while, in non-normalized Euclidean distance, the amplitude matters. Ultimately, the original matrix profile researchers, who have thought about this stuff way more than I have, have found z-normalization to be generally applicable and that it is now fast to compute all of the pairwise distances with it. With other distances, these facts may not hold true.

**Reply by hugo-pires** [reply]
Thank you. I don't have any case yet but I thought that, as you may consider the window length as a parameter, the Minkowski distance `p` exponent could be also considered. But It is just a guess and I am not a specialist on the matter.

**Reply by seanlaw** [reply]
> Thank you. I don't have any case yet but I thought that, as you may consider the window length as a parameter, the Minkowski distance `p` exponent could be also considered. But It is just a guess and I am not a specialist on the matter.

So, adding Minkowski distance has been [requested in the past](https://github.com/TDAmeritrade/stumpy/issues/251)  and this feature was successfully added to the main branch just two days ago:

https://github.com/TDAmeritrade/stumpy/commit/d3b68b4daf0fa76697c7b6aa17adb02591a75388

And it should be available in the next STUMPY release. So, you will be able to do something like:

```
import numpy as np
import stumpy

T = np.random.rand(100)
m = 5
mp = stumpy.stump(T, m, normalize=False, p=1.0)
```

And this would compute the matrix profiles using Manhattan distance rather than Euclidean distance (without z-normalization). If you'd like to clone the repo and test it out, I'd appreciate any feedback/criticism!
